File: Problem/By section/Tree/Problem 144/problem_144.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
**/
struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode() : val(0), left(nullptr), right(nullptr) {}
     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    // Time: O(n), Space: O(n)
    vector<int> preorderTraversal(TreeNode* root) {
        if(!root) return vector<int>();
        stack<TreeNode*> s;
        vector<int> result;
        s.push(root);
        
        while(!s.empty()) {
            auto node = s.top();
            s.pop();
            
            result.push_back(node->val);
            if(node->right) s.push(node->right);
            if(node->left) s.push(node->left);
        }
        
        return result;
    }
};

int main(int argc, char** argv) {
    Solution s;
    
    return 0;
}

File: Problem/By section/Tree/Problem 144/problem_144.java
import java.util.*;


class Solution {
    
//     public List<Integer> preorderTraversal(TreeNode root) {
//         List<Integer> ans = new LinkedList<>();
//         dfs(root, ans);
//         return ans;
//     }
    
//     public void dfs(TreeNode root, List<Integer> ans){
//         if(root == null) return;
        
//         ans.add(root.val);
//         dfs(root.left, ans);
//         dfs(root.right, ans);
//     }
    
    public List<Integer> preorderTraversal(TreeNode root) {
        if(root == null) return new LinkedList<>();;

        List<Integer> ans = new LinkedList<>();
        Stack<TreeNode> stack = new Stack<>();
        stack.add(root);

        while(!stack.isEmpty()){
            TreeNode cur_Node = stack.pop();
            ans.add(cur_Node.val);
            
            if(cur_Node.right != null) stack.add(cur_Node.right);
            if(cur_Node.left != null) stack.add(cur_Node.left);
        }

        return ans;
    }
}